Adds boolean to Markdown for HTML escaping

edwardreed81 opened the pull request:
This allows the developer to render raw HTML inside Markdown if desired.

valentijnnieman commented:
Thanks for making this PR! This looks good to me.

@chriddyp Is there any reason (security?) why the Markdown component doesn't support the `escapeHtml` prop already?

chriddyp commented:
Yeah, we might want to do a quick rename on this prop, like `dangerouslyEscapeHtml`, similar to https://github.com/plotly/dash-dangerously-set-inner-html

chriddyp commented:
and our docstring for the prop should describe why this can be dangerous as well

edwardreed81 commented:
I'm happy to rename the prop to remind people it is dangerous, but I don't feel qualified to issue warnings about the dangerousness of this... someone else want to take a shot at it?

valentijnnieman commented on package.json:
```diff
@@ -50,7 +50,7 @@
     "react-docgen": "^2.20.1",
     "react-dropzone": "^4.1.2",
-    "react-markdown": "^2.4.5",
+    "react-markdown": "^4.0",
```
Could you remove this line in favour of doing it in a separate PR, possibly #394?

edwardreed81 replied:
Whoops, I got my branches mixed up apparently. This'll be reverted in the next commit.

valentijnnieman commented on src/components/Markdown.react.js:
```diff
@@ -13,7 +13,7 @@ function DashMarkdown(props) {
     }
 
-    return <Markdown source={props.children} escapeHtml={true} {...props} />;
+    return <Markdown source={props.children} {...props} />;
```
I think you're missing something like `escapeHtml={props.dangerouslyEscapeHtml}` here - the `{...props}` spread here isn't enough because of the naming difference (dangerouslyEscapeHtml vs escapeHtml). Could you add that?

edwardreed81 replied:
Thanks for pointing this out, I forgot. Once I resolve the naming/boolean question below, it'll be in the next commit.

valentijnnieman commented on src/components/Markdown.react.js:
```diff
@@ -29,6 +29,14 @@ DashMarkdown.propTypes = {
+     * inadvertently expose your users to a cross-site scripting (XSS)
+     * (https://en.wikipedia.org/wiki/Cross-site_scripting) attack.
+     */
+    dangerouslyEscapeHtml: PropTypes.bool,
```
👍

edwardreed81 commented:
I just remembered that `escapeHtml = true` for `react-markdown` is the safe choice; rendering raw HTML requites `escapeHtml = false`.

I could make the option inside `dcc` something like `dangerouslyRenderHtml` and have the boolean values reversed from how `react-markdown` handles it. Or keep paradigm and name the parameter `safelyEscapeHtml`.

What do you guys think?

valentijnnieman commented:
In that case, I think it would be best to stick to the naming `react-markdown` uses. That way, the prop in Dash maps 1-to-1 to `react-markdown`.

edwardreed81 commented:
Do you mean just use `escapeHtml` like `react-markdown` or use `safelyEscapeHtml`?

edwardreed81 commented:
I reverted back to using just `escapeHtml` to match what `react-markdown` does. I added a default value of `true` for this prop (the safe choice).

Not sure why the checks failed; I don't know what `prettier` is.

chriddyp commented:
I would still prefer to warn users that providing raw HTML is dangerous. For many dash users, this will be their first encounter with XSS-like attacks, and I'd like to be very upfront about how dangerous this can be.

Keeping the props matching 1-1 is nice, but it's not required by us. There is already a precedent for us _not_ doing this, e.g. with the `DatePickerSingle`, we convert the prop types to underscore-case.

So, I'd say we name the prop something like `dangerously_allow_html` (which gets mapped as `escapeHtml = !dangerously_allow_html`). This is similar to the language that we use elsewhere ("dangerouslySetInnerHtml")

edwardreed81 commented:
I'm happy with that @chriddyp. I made those changes, but again I don't understand why the checks are failing.

edwardreed81 commented:
What needs to happen to get this merged?

edwardreed81 commented:
@valentijnnieman the Percy check fails now, but I get "Something went wrong" when I click Details. Any idea?

valentijnnieman commented:
@edwardreed81 Ah that's probably because you don't have access. I've taken a look and looks like it's all good - Percy takes snapshots of what it thinks are changes, but it all looks good to me.
